scripts/o3de/o3de/compatibility.py

```diff
@@ -0,0 +1,295 @@
+    if not project_json_data:
+        logger.error(f'Failed to load project.json data from {project_path} needed for checking compatibility')
+        return False
+
+    engine_name = project_json_data.get('engine','')
+    if engine_name:
+        engine_json_data = manifest.get_engine_json_data(engine_name=engine_name)
+    else:
+        engine_json_data = manifest.get_engine_json_data(engine_path=manifest.get_this_engine_path())
+    if not engine_json_data:
+        logger.error(f'Failed to load engine.json data based on the engine field in project.json or detect the engine from the current folder')
+        return False
+
+    # compatibility will be based on the engine the project uses and the gems visible to
+    # the engine and project
+    compatible = gem_engine_compatible(gem_json_data, engine_json_data, gem_paths=gem_paths)
+    if compatible and check:
+        logger.info(f'{gem_json_data["gem_name"]} is compatible with this project and engine.')
+    elif not compatible:
+        return False
+
+    return True 
+
+
+def gem_engine_compatible(gem_json_data:dict, engine_json_data:dict, gem_paths:list = None, check:bool = False) -> bool:
+    """
+    Returns True if the gem is compatible with the indicated engine.
+    :param gem_json_data: gem json data dictionary
+    :param engine_json_data: engine json data dictionary
+    :param gem_paths: paths to all gems to use for dependency checks (optional)
+    :param check: if True always return True and ouput log info about the result
+    """
+    engine_compatible = o3de_object_is_compatible_with_engine(gem_json_data, engine_json_data)
+    if not engine_compatible:
+        error_msg = f'{gem_json_data["gem_name"]} is not known to be compatible with the '\
+            f'engine {engine_json_data["engine_name"]}{engine_json_data.get("version","")} '\
+             'and requires the --force parameter to enable.'
+        if check:
+            logger.info(error_msg)
+        else:
+            logger.error(error_msg)
+
+    if not gem_paths:
+        gem_paths = manifest.get_engine_gems()
+
+    return engine_compatible and gem_dependencies_compatible(gem_json_data, gem_paths=gem_paths, check=check)
+
+def o3de_object_is_compatible_with_engine(object_json_data:dict, engine_json_data:dict) -> bool:
+    """
+    Returns True if the engine is compatible with the provided object's compatible_engines and engine_api_versions information.
+    :param object_json_data: the json data for the object 
+    :param engine_json_data: the json data for the engine
+    """
+    if (object_json_data.get('compatible_engines') or object_json_data.get('engine_api_dependencies')):
+        return engine_is_compatible(
+            engine_json_data['engine_name'], engine_json_data.get('version',''), 
+            object_json_data.get('compatible_engines',''), object_json_data.get('engine_api_dependencies',''),
+            engine_json_data)
+    return True
```

Comment:
This can be a single return statement. Since it is only boolean checks.
```suggestion
    return object_json_data.get('compatible_engines') or object_json_data.get('engine_api_dependencies') \
        engine_is_compatible(
            engine_json_data['engine_name'], engine_json_data.get('version',''), 
            object_json_data.get('compatible_engines',''), object_json_data.get('engine_api_dependencies',''),
            engine_json_data)
```